app/src/main/java/org/kiwix/kiwixmobile/zim_manager/local_file_transfer/LocalFileTransferActivity.java

```diff
@@ -0,0 +1,437 @@
+        case PERMISSION_REQUEST_CODE_COARSE_LOCATION: {
+          Log.e(TAG, "Location permission not granted");
+
+          showToast(this, R.string.permission_refused_location, Toast.LENGTH_LONG);
+          finish();
+          break;
+        }
+
+        case PERMISSION_REQUEST_CODE_STORAGE_WRITE_ACCESS: {
+          Log.e(TAG, "Storage write permission not granted");
+
+          showToast(this, R.string.permission_refused_storage, Toast.LENGTH_LONG);
+          finish();
+          break;
+        }
+
+        default: {
+          super.onRequestPermissionsResult(requestCode, permissions, grantResults);
+          break;
+        }
+      }
+    }
+  }
+
+  private boolean isLocationServicesEnabled() {
+    LocationManager locationManager =
+        (LocationManager) getApplicationContext().getSystemService(Context.LOCATION_SERVICE);
+    boolean gps_enabled = false;
+    boolean network_enabled = false;
+
+    try {
+      gps_enabled = locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER);
+    } catch (Exception ex) {
+      ex.printStackTrace();
+    }
+
+    try {
+      network_enabled = locationManager.isProviderEnabled(LocationManager.NETWORK_PROVIDER);
+    } catch (Exception ex) {
+      ex.printStackTrace();
+    }
+
+    return (gps_enabled || network_enabled);
+  }
+
+  private void requestEnableLocationServices() {
+    alertDialogShower.show(KiwixDialog.EnableLocationServices.INSTANCE,
+        new Function0<Unit>() {
+          @Override public Unit invoke() {
+            startActivityForResult(new Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS),
+                REQUEST_ENABLE_LOCATION_SERVICES);
+            return Unit.INSTANCE;
+          }
+        },
+        new Function0<Unit>() {
+          @Override public Unit invoke() {
+            showToast(LocalFileTransferActivity.this, R.string.discovery_needs_location,
+                Toast.LENGTH_SHORT);
+            return Unit.INSTANCE;
+          }
+        });
+  }
+
+  private void requestEnableWifiP2pServices() {
+    alertDialogShower.show(KiwixDialog.EnableWifiP2pServices.INSTANCE,
+        new Function0<Unit>() {
+          @Override public Unit invoke() {
+            startActivity(new Intent(Settings.ACTION_WIFI_SETTINGS));
+            return Unit.INSTANCE;
+          }
+        },
+        new Function0<Unit>() {
+          @Override public Unit invoke() {
+            showToast(LocalFileTransferActivity.this, R.string.discovery_needs_wifi,
+                Toast.LENGTH_SHORT);
+            return Unit.INSTANCE;
+          }
+        });
+  }
+
+  @Override
+  protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
+    switch (requestCode) {
+      case REQUEST_ENABLE_LOCATION_SERVICES: {
+        LocationManager locationManager =
+            (LocationManager) getApplicationContext().getSystemService(Context.LOCATION_SERVICE);
+
+        if (!(locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER)
```

Comment:
you don't check exceptions here? Is the exception checking earlier really necessary? Also couldn't you just use that earlier method?

Reply:
iirc I was using the code directly for running the program initially. I added checks for permissions and the method later, so I must have forgotten to use it everywhere
[Done](https://github.com/kiwix/kiwix-android/pull/1276/commits/cb944bb88f3809b3baee4613e21e92972122a775#diff-e1d59ae6e15dcb502e125fb6e243d15dR386) now

Is the switch alright here (single case only as of now) or should I convert it to an `if`?

Reply:
ehhh some codestyles get really snitty about this. I don't care too much and I kind of expect `onActivityResult`s to look like this. It is an autocomplete action to convert switch to if and vice versa. It is more concise to be an if now and more capable of being extended as a switch.
The kotlin [when](https://kotlinlang.org/docs/reference/control-flow.html#when-expression) is just so superior, no stupid breaks.
Tl:dr do whatever you like

Reply:
I think I'll leave it as such, just in case android decides to require more permissions for p2p in the future
Also kotlin syntax seems to avoid boiler plate even more than a couple of general purpose languages that I have tried - looking forward to learning kt in the coming months